src/monitors/BytecodeProfilingMonitor.v3

```diff
@@ -0,0 +1,450 @@
+// Copyright 2023 Wizard Authors. All rights reserved.
+// See LICENSE for details of Apache 2.0 license.
+
+// Implements a profiler that targets wasm modules containing a generalized interpreter
+// that evaluates compiled bytecode.
+//
+// This profiler aims to recover source level information, including hot functions, call graphs,
+// and execution costs in terms of wasm instructions.
+def E = MonitorRegistry.add(
+	"profile_bytecode", "Recovers source level information from bytecode execution",
+	BytecodeProfilingMonitor.new());
+
+def OUT = Trace.OUT;
+
+// the minimum number of cases contained in the "switch" statement of the dispatch
+// function. This is used during static analysis to identify the most likely wasm function
+// as dispatch.
+var MIN_DISPATCH_CASE_COUNT = 100;
+
+// the maximum size in bytes that separates 2 consecutive opcodes.
+var MAX_OPERAND_LEN = 6;
+
+class BytecodeProfilingMonitor extends Monitor {
+	// all functions that could be the dispatch function.
+	// kvp of function index to the (FuncDecl, bytecode_read_pc, profile_state) pair.
+	var potential_dispatch_functions = HashMap<int, (FuncDecl, int, SourceProfileState)>.new(int.!<int>, int.==);
+	var all_counters: List<DynamicLoc -> Resumption>;
+
+	def configure(args: string) -> string {
+		def argMap = MonitorUtil.argMapFrom(args);
+		if (argMap.has("switch_size")) {
+			var parse_result = Ints.parseDecimal(argMap["switch_size"], 0);
+			if (parse_result.0 > 0) MIN_DISPATCH_CASE_COUNT = parse_result.1;
+			return null;
+		}
+		if (argMap.has("operand_len")) {
+			var parse_result = Ints.parseDecimal(argMap["operand_len"], 0);
+			if (parse_result.0 > 0) MAX_OPERAND_LEN = parse_result.1;
+			return null;
+		}
+		def buf = StringBuilder.new();
+		buf.put1("invalid argument(s) \"%s\"", args);
+		return buf.toString();
+	}
+
+	def onParse(module: Module) {
+		var mi = ModuleInstrumenter.new(module);
+
+		var bi = BytecodeIterator.new();
+		for (i < module.functions.length) {
+			var f = module.functions[i];
+			if (f.imp != null) continue;
+			bi = bi.reset(f);
+
+			// We identify dispatch function through static analysis, by looking for:
+			// 	1. a sequence of consequtive BLOCK instructions exceeding MIN_DISPATCH_CASE_COUNT.
+			// 	   This corresponds to a switch statement over all possible opcode in the source code.
+			// 	2. a I32_LOAD8_U instruction, followed by a BR_TABLE instruction, in the inner scope of the
+			// 	   nested BLOCK chain.
+			var prev_OPCODE_READ_pc = -1, bytecode_read_pc = -1;
+			var last_block_chain = 0;
```

Comment:
What about instead just matching on the number of cases in a `br_table`? That way you don't have to track the number of consecutive blocks, but it'd rather be determined by how many cases there are (though a `br_table` might have many redundant entries).

Reply:
Good idea, will make the change